src/java.base/aix/native/libjava/ProcessHandleImpl_aix.c

```diff
@@ -36,9 +38,127 @@
 
 void os_initNative(JNIEnv *env, jclass clazz) {}
 
+/*
+ * Returns the children of the requested pid and optionally each parent and
+ * start time. If requested pid is zero return all processes.
+ * Use getprocs64 to accumulate any process following the rule above.
+ * The resulting pids are stored into the array of longs.
+ * The number of pids is returned if they all fit.
+ * If the parentArray is non-null, store the parent pid.
+ * If the array is too short, excess pids are not stored and
+ * the desired length is returned.
+ */
 jint os_getChildren(JNIEnv *env, jlong jpid, jlongArray jarray,
                     jlongArray jparentArray, jlongArray jstimesArray) {
-    return unix_getChildren(env, jpid, jarray, jparentArray, jstimesArray);
+    pid_t pid = (pid_t) jpid;
+    jlong* pids = NULL;
+    jlong* ppids = NULL;
+    jlong* stimes = NULL;
+    jsize parentArraySize = 0;
+    jsize arraySize = 0;
+    jsize stimesSize = 0;
+    jsize count = 0;
+
+    arraySize = (*env)->GetArrayLength(env, jarray);
+    JNU_CHECK_EXCEPTION_RETURN(env, -1);
+    if (jparentArray != NULL) {
+        parentArraySize = (*env)->GetArrayLength(env, jparentArray);
+        JNU_CHECK_EXCEPTION_RETURN(env, -1);
+
+        if (arraySize != parentArraySize) {
+            JNU_ThrowIllegalArgumentException(env, "array sizes not equal");
+            return 0;
+        }
+    }
+    if (jstimesArray != NULL) {
+        stimesSize = (*env)->GetArrayLength(env, jstimesArray);
+        JNU_CHECK_EXCEPTION_RETURN(env, -1);
+
+        if (arraySize != stimesSize) {
+            JNU_ThrowIllegalArgumentException(env, "array sizes not equal");
+            return 0;
+        }
+    }
+
+    const int chunk = 100;
+    struct procentry64 ProcessBuffer[chunk];
+    pid_t idxptr = 0;
+    int i, num = 0;
+
+    do { // Block to break out of on Exception
+        pids = (*env)->GetLongArrayElements(env, jarray, NULL);
```

Comment:
Nit, I'd move these invocations of GLAE into the initialization block and remove the outer loop. I see what you do here, the outer block gives you a reliable jumping point to get cleanup in case of an error. But I would just use a goto cleanup label. That's clearer and allows you to handle initialization in one place.

Proposal for initialization:

```
if (jparentArray != null) {
  - check size is same as primary array, if wrong throw + goto cleanup
  - ppids = GLAE
} else {
  - if input pid is 0, throw an error too, since for "get all processes mode"
      we need the parent array. Throw + goto cleanup.
}
// same for times array
```

Reply:
hmm? As I already mentioned I do not want to change the structure of the function with this pr. This can be done by someone else for macos, linux and aix in parallel.

Reply:
If someone reworks the code, you would have to counter-check the code anyway since nobody else has AIX machines around to build and test. So I'd just do it to save some cycles. But I leave it up to you.